src/Generating/FinishGen.cpp

```diff
@@ -252,20 +252,20 @@ void cFinishGenClumpTopBlock::TryPlaceFoliageClump(cChunkDesc & a_ChunkDesc, int
 		int z = a_CenterZ + (((rnd / 256) % RANGE_FROM_CENTER * 2) - RANGE_FROM_CENTER);
 		int Top = a_ChunkDesc.GetHeight(x, z);
 
-		if (a_ChunkDesc.GetBlockType(x, Top, z) != E_BLOCK_GRASS)
+		auto GroundBlockType = a_ChunkDesc.GetBlockType(x, Top, z);
+		if ((GroundBlockType == E_BLOCK_GRASS) ||
+			((GroundBlockType == E_BLOCK_MYCELIUM) && ((a_BlockType == E_BLOCK_RED_MUSHROOM) || (a_BlockType == E_BLOCK_BROWN_MUSHROOM))))
 		{
-			continue;
-		}
-
-		a_ChunkDesc.SetBlockTypeMeta(x, Top + 1, z, a_BlockType, a_BlockMeta);
-		if (a_IsDoubleTall)
-		{
-			a_ChunkDesc.SetBlockTypeMeta(x, Top + 2, z, E_BLOCK_BIG_FLOWER, E_META_BIG_FLOWER_TOP);
-			a_ChunkDesc.SetHeight(x, z, static_cast<HEIGHTTYPE>(Top + 2));
-		}
-		else
-		{
-			a_ChunkDesc.SetHeight(x, z, static_cast<HEIGHTTYPE>(Top + 1));
+			a_ChunkDesc.SetBlockTypeMeta(x, Top + 1, z, a_BlockType, a_BlockMeta);
+			if (a_IsDoubleTall)
+			{
+				a_ChunkDesc.SetBlockTypeMeta(x, Top + 2, z, E_BLOCK_BIG_FLOWER, E_META_BIG_FLOWER_TOP);
```

Comment:
Just to confirm, mycelium mushrooms are never a_IsDoubleTall, right?

Reply:
Yes the code checks when calling ```TryPlaceFoliageClump``` if the block is a Tall flower: 
```TryPlaceFoliageClump(a_ChunkDesc, PosX, PosZ, Block.m_BlockType, Block.m_BlockMeta, Block.m_BlockType == E_BLOCK_BIG_FLOWER);```